// twelvelabs-js/dist/resources/video/index.mjs
import * as Models from "../../models/index.mjs";
import { APIResource } from "../../resource.mjs";
import { convertKeysToSnakeCase, handleComparisonParams, removeUndefinedValues } from "../../util.mjs";
export class Video extends APIResource {
    async retrieve(indexId, id, { embed } = {}, options = {}) {
        const _params = convertKeysToSnakeCase({ embed });
        const res = await this._get(`indexes/${indexId}/videos/${id}`, removeUndefinedValues(_params), { ...options, skipCamelKeys: ['metadata'] });
        return new Models.Video(this, indexId, res);
    }
    async list(indexId, { size, width, height, duration, fps, createdAt, updatedAt, indexedAt, ...restParams } = {}, options = {}) {
        const _params = convertKeysToSnakeCase(restParams);
        handleComparisonParams(_params, 'size', size);
        handleComparisonParams(_params, 'width', width);
        handleComparisonParams(_params, 'height', height);
        handleComparisonParams(_params, 'duration', duration);
        handleComparisonParams(_params, 'fps', fps);
        handleComparisonParams(_params, 'createdAt', createdAt);
        handleComparisonParams(_params, 'updatedAt', updatedAt);
        handleComparisonParams(_params, 'indexedAt', indexedAt);
        const res = await this._get(`indexes/${indexId}/videos`, removeUndefinedValues(_params), { ...options, skipCamelKeys: ['user_metadata'] });
        return res.data.map((v) => new Models.Video(this, indexId, v));
    }
    async listPagination(indexId, { size, width, height, duration, fps, createdAt, updatedAt, indexedAt, ...restParams } = {}, options = {}) {
        const originParams = {
            size,
            width,
            height,
            duration,
            fps,
            createdAt,
            updatedAt,
            indexedAt,
            ...restParams,
        };
        const _params = convertKeysToSnakeCase(restParams);
        handleComparisonParams(_params, 'size', size);
        handleComparisonParams(_params, 'width', width);
        handleComparisonParams(_params, 'height', height);
        handleComparisonParams(_params, 'duration', duration);
        handleComparisonParams(_params, 'fps', fps);
        handleComparisonParams(_params, 'createdAt', createdAt);
        handleComparisonParams(_params, 'updatedAt', updatedAt);
        handleComparisonParams(_params, 'indexedAt', indexedAt);
        const res = await this._get(`indexes/${indexId}/videos`, removeUndefinedValues(_params), { ...options, skipCamelKeys: ['user_metadata'] });
        return new Models.VideoListWithPagination(this, originParams, indexId, res.data, res.pageInfo);
    }
    async update(indexId, id, { userMetadata }, options = {}) {
        await this._put(`indexes/${indexId}/videos/${id}`, removeUndefinedValues(convertKeysToSnakeCase({ userMetadata })), options);
    }
    async delete(indexId, id, options = {}) {
        await this._delete(`indexes/${indexId}/videos/${id}`, options);
    }
}
