// twelvelabs-js/dist/resources/task/index.mjs
import * as Models from "../../models/index.mjs";
import { attachFormFile, convertKeysToSnakeCase, FormDataImpl, handleComparisonParams, removeUndefinedValues, } from "../../util.mjs";
import { APIResource } from "../../resource.mjs";
export class TaskTransfer extends APIResource {
    async importVideos({ integrationId, ...restParams }, options = {}) {
        const _params = convertKeysToSnakeCase(restParams);
        const res = await this._post(`tasks/transfers/import/${integrationId}`, removeUndefinedValues(_params), options);
        return res;
    }
    async importStatus(integrationId, indexId, options = {}) {
        const _params = convertKeysToSnakeCase({ indexId });
        const res = await this._get(`tasks/transfers/import/${integrationId}/status`, removeUndefinedValues(_params), options);
        return res;
    }
    async importLogs(integrationId, options = {}) {
        const res = await this._get(`tasks/transfers/import/${integrationId}/logs`, {}, options);
        return res.data || [];
    }
}
export class Task extends APIResource {
    constructor(client) {
        super(client);
        this.transfers = new TaskTransfer(client);
    }
    async retrieve(id, options = {}) {
        const res = await this._get(`tasks/${id}`, {}, options);
        return new Models.Task(this, res);
    }
    async list({ createdAt, updatedAt, ...restParams } = {}, options = {}) {
        const _params = convertKeysToSnakeCase(restParams);
        handleComparisonParams(_params, 'createdAt', createdAt);
        handleComparisonParams(_params, 'updatedAt', updatedAt);
        const res = await this._get('tasks', removeUndefinedValues(_params), options);
        return res.data.map((v) => new Models.Task(this, v));
    }
    async listPagination({ createdAt, updatedAt, ...restParams } = {}, options = {}) {
        const originParams = { createdAt, updatedAt, ...restParams };
        const _params = convertKeysToSnakeCase(restParams);
        handleComparisonParams(_params, 'createdAt', createdAt);
        handleComparisonParams(_params, 'updatedAt', updatedAt);
        const res = await this._get('tasks', removeUndefinedValues(_params), options);
        return new Models.TaskListWithPagination(this, originParams, res.data, res.pageInfo);
    }
    async create(body, options = {}) {
        if (!body.file && !body.url) {
            throw new Error('Either file or url must be provided');
        }
        const formData = new FormDataImpl();
        formData.append('index_id', body.indexId);
        if (body.url)
            formData.append('video_url', body.url);
        if (body.transcriptionUrl)
            formData.append('transcription_url', body.transcriptionUrl);
        if (body.language)
            formData.append('language', body.language);
        if (body.enableVideoStream)
            formData.append('enable_video_stream', String(body.enableVideoStream));
        try {
            if (body.file)
                attachFormFile(formData, 'video_file', body.file);
            if (body.transcriptionFile) {
                attachFormFile(formData, 'transcription_file', body.transcriptionFile);
                formData.append('provide_transcription', 'true');
            }
        }
        catch (err) {
            throw err;
        }
        const res = await this._post('tasks', formData, options);
        return await this.retrieve(res.id);
    }
    async createBulk(indexId, { files, urls, language, enableVideoStream, }, options = {}) {
        if (!files && !urls) {
            throw new Error('Either files or urls must be provided');
        }
        const tasks = [];
        if (files) {
            for (const file of files) {
                try {
                    const task = await this.create({ indexId, file, language, enableVideoStream }, options);
                    tasks.push(task);
                }
                catch (e) {
                    console.error(`Error processing file ${file}:`, e);
                }
            }
        }
        if (urls) {
            for (const url of urls) {
                try {
                    const task = await this.create({ indexId, url, language, enableVideoStream }, options);
                    tasks.push(task);
                }
                catch (e) {
                    console.error(`Error processing url ${url}:`, e);
                }
            }
        }
        return tasks;
    }
    async delete(id, options = {}) {
        await this._delete(`tasks/${id}`, options);
    }
}
