// twelvelabs-js/dist/resources/generate/index.mjs
import * as Models from "../../models/index.mjs";
import { APIResource } from "../../resource.mjs";
import { convertKeysToSnakeCase, removeUndefinedValues, trackStream } from "../../util.mjs";
export class Generate extends APIResource {
    async summarize(videoId, type, prompt, temperature, options = {}) {
        const _body = convertKeysToSnakeCase({
            videoId,
            type,
            prompt,
            temperature,
        });
        const res = await this._post('summarize', removeUndefinedValues(_body), options);
        return res;
    }
    async gist(videoId, types, options = {}) {
        const _body = convertKeysToSnakeCase({
            videoId,
            types,
        });
        const res = await this._post('gist', _body, options);
        return res;
    }
    async text(videoId, prompt, temperature, options = {}) {
        const _body = convertKeysToSnakeCase({
            videoId,
            prompt,
            temperature,
            stream: false,
        });
        const res = await this._post('generate', removeUndefinedValues(_body), options);
        return res;
    }
    async textStream({ videoId, prompt, temperature }, options = {}) {
        const _body = convertKeysToSnakeCase({
            videoId,
            prompt,
            temperature,
            stream: true,
        });
        const res = await this._post('generate', removeUndefinedValues(_body), options);
        return new Models.GenerateTextStreamResult(trackStream(res));
    }
}
