// twelvelabs-js/dist/resources/embed/index.mjs
import FormData from 'form-data';
import * as Models from "../../models/index.mjs";
import { APIResource } from "../../resource.mjs";
import { attachFormFile, removeUndefinedValues, convertKeysToSnakeCase } from "../../util.mjs";
export class EmbedTask extends APIResource {
    async retrieve(id, options = {}) {
        const res = await this._get(`embed/tasks/${id}`, {}, options);
        return new Models.EmbeddingsTask(this, res);
    }
    async list(params = {}, options = {}) {
        const _params = convertKeysToSnakeCase(params);
        const res = await this._get('embed/tasks', removeUndefinedValues(_params), options);
        return res.data.map((v) => new Models.EmbeddingsTask(this, v));
    }
    async listPagination(params = {}, options = {}) {
        const _params = convertKeysToSnakeCase(params);
        const res = await this._get('embed/tasks', removeUndefinedValues(_params), options);
        return new Models.EmbeddingsTaskListWithPagination(this, params, res.data, res.pageInfo);
    }
    async create(modelName, { file, url, startOffsetSec, endOffsetSec, clipLength, scopes }, options = {}) {
        if (!file && !url) {
            throw new Error('Either video file or url must be provided');
        }
        const formData = new FormData();
        formData.append('model_name', modelName);
        if (url)
            formData.append('video_url', url);
        if (startOffsetSec)
            formData.append('video_start_offset_sec', startOffsetSec);
        if (endOffsetSec)
            formData.append('video_end_offset_sec', endOffsetSec);
        if (clipLength)
            formData.append('video_clip_length', clipLength);
        if (scopes) {
            scopes.forEach((scope) => formData.append('video_embedding_scope', scope));
        }
        try {
            if (file)
                attachFormFile(formData, 'video_file', file);
        }
        catch (err) {
            throw err;
        }
        const { id } = await this._post('embed/tasks', formData, options);
        const task = await this.retrieve(id);
        return task;
    }
    async createBulk(modelName, videos, options = {}) {
        const tasks = [];
        for (const videoParams of videos) {
            try {
                const task = await this.create(modelName, videoParams, options);
                tasks.push(task);
            }
            catch (e) {
                console.error(`Error creating task with video: ${e}`);
            }
        }
        return tasks;
    }
    async status(taskId, options = {}) {
        const res = await this._get(`embed/tasks/${taskId}/status`, {}, options);
        return new Models.EmbeddingsTaskStatus(res);
    }
}
export class Embed extends APIResource {
    constructor(client) {
        super(client);
        this.task = new EmbedTask(client);
    }
    async create({ modelName, text, textTruncate, audioUrl, audioFile, audioStartOffsetSec, imageUrl, imageFile, }, options = {}) {
        if (!text && !audioUrl && !audioFile && !imageUrl && !imageFile) {
            throw new Error('At least one of text, audioUrl, audioFile, imageUrl, imageFile must be provided');
        }
        const formData = new FormData();
        formData.append('model_name', modelName);
        if (text)
            formData.append('text', text);
        if (textTruncate)
            formData.append('text_truncate', textTruncate);
        if (audioUrl)
            formData.append('audio_url', audioUrl);
        if (imageUrl)
            formData.append('image_url', imageUrl);
        if (audioFile)
            attachFormFile(formData, 'audio_file', audioFile);
        if (imageFile)
            attachFormFile(formData, 'image_file', imageFile);
        if (audioStartOffsetSec)
            formData.append('audio_start_offset_sec', audioStartOffsetSec);
        const res = await this._post('embed', formData, options);
        return new Models.CreateEmbeddingsResult(res);
    }
}
